namui-cli/src/services/rust_build_service.rs

```diff
@@ -0,0 +1,266 @@
+    },
+    thread,
+    time::Duration,
+};
+
+pub struct RustBuildService {
+    builder: Mutex<Option<Arc<CancelableBuilder>>>,
+    is_build_just_started: AtomicBool,
+}
+pub enum BuildResult {
+    Canceled,
+    Successful(CargoBuildResult),
+    Failed(String), // It's not failed if cargo build result has error messages.
+}
+
+#[derive(Clone)]
+pub struct BuildOption {
+    pub dist_path: PathBuf,
+    pub project_root_path: PathBuf,
+    pub platform: BuildPlatform,
+}
+
+#[derive(Clone)]
+pub enum BuildPlatform {
+    WasmWeb,
+}
+
+impl RustBuildService {
+    pub(crate) fn new() -> Self {
+        Self {
+            builder: Mutex::new(None),
+            is_build_just_started: AtomicBool::new(false),
+        }
+    }
+
+    pub(crate) fn cancel_and_start_build(&self, build_option: &BuildOption) -> BuildResult {
+        if self.is_build_just_started.swap(true, Ordering::Relaxed) {
+            return BuildResult::Canceled;
+        }
+
+        let result_receiver = {
+            let mut builder_lock = self.builder.lock().unwrap();
+            if let Some(builder) = builder_lock.take() {
+                builder.cancel();
+            }
+
+            self.is_build_just_started.store(false, Ordering::Relaxed);
+            println!("starting build");
+            let (builder, result_receiver) = CancelableBuilder::start(build_option);
+            *builder_lock = Some(builder);
+            result_receiver
+        };
+
+        result_receiver.recv().unwrap()
+    }
+}
+
+struct CancelableBuilder {
+    is_cancel_requested: AtomicBool,
+    is_canceled: AtomicBool,
+}
+
+impl CancelableBuilder {
+    pub fn cancel(&self) {
+        if self.is_canceled.load(Ordering::Relaxed) {
+            return;
+        }
+
+        self.is_cancel_requested.store(true, Ordering::Relaxed);
+        debug_println!("build cancel requested");
+
+        loop {
+            thread::sleep(Duration::from_millis(100));
+
+            if self.is_canceled.load(Ordering::Relaxed) {
+                break;
+            }
+        }
+    }
+
+    pub fn start(
+        build_option: &BuildOption,
+    ) -> (
+        Arc<CancelableBuilder>,
+        std::sync::mpsc::Receiver<BuildResult>,
+    ) {
+        let builder = Arc::new(Self {
+            is_cancel_requested: AtomicBool::new(false),
+            is_canceled: AtomicBool::new(false),
+        });
+        let build_option = build_option.clone();
+
+        let builder_thread_fn = {
+            move |builder: Arc<Self>| -> Result<BuildResult, Box<dyn std::error::Error>> {
+                let mut spawned_process = Self::spawn_build_process(&build_option)?;
+
+                let mut stdout = spawned_process.stdout.take().unwrap();
+                let mut stderr = spawned_process.stderr.take().unwrap();
+                let stdout_reading_thread = thread::spawn(move || {
+                    let mut string = String::new();
+                    stdout.read_to_string(&mut string).unwrap();
```

Comment:
there are many report about stuck the child process when they didn't flush outputs. it is blocking function so I ran thread for it.

ref: https://stackoverflow.com/questions/34611742/how-do-i-read-the-output-of-a-child-process-without-blocking-in-rust